fundamentals/computing-development-environments/docker.md

```diff
@@ -1,3 +1,127 @@
+
+- A server with a long-running daemon process `dockerd`.
+- APIs which specify interfaces that programs can use to talk to and instruct the Docker daemon.
+- A command line interface (CLI) client `docker`.
+
+You can read more details about the Docker Architecture [here](https://docs.docker.com/get-started/overview/#docker-architecture).
+
+We'll cover **Registry** on the core concept below.
+
+### Docker Installation
+
+You can install docker directly from their website at [https://docs.docker.com/get-docker/](https://docs.docker.com/get-docker/).
+
+## Concepts of Docker
+
+### 1. Docker file
+
+A Docker file is a file containing steps to building your docker image.
+
+Example of an awesome `Dockerfile` can be found [here](https://github.com/pangeo-data/pangeo-docker-images/blob/master/base-image/Dockerfile). What's great about this image is that it sets up conda and jupyter server, which is useful when working on research related work.
+
+With Docker file, you can share this with anyone and that person will get the same exact image as you after it's built.
+Also, since it is a file, you can version control it with git.
+
+### 2. Docker Image
+
+A Docker Image is the "built" version of a Docker file.
+This typically live on your internal Docker host or an external Registry, so you can't store a Docker image in git.
+
+An example of a docker image that has been built with the Docker file above can be found [here](https://hub.docker.com/layers/pangeo/base-image/latest/images/sha256-1854ca62ef75f1e017e1920b3a167c62cb5f0ee921f13c3247fc4b016e5be3be?context=explore). That is the published image for with specific file sha, which allows for a specific version of that image, just like your regular files!
+
+### 3. Docker Container
+
+A container is a standard unit of software that get's spun up from an image.
+This is the running software application that you have built.
+It is lightweight and easily scalable and tools like [Kubernetes](https://kubernetes.io/) are used to orchestrate the deployment of these containers in production.
+
+### 4. Docker Orchestration
+
+Containerization has changed how we build and run applications. It allows apps to run consistently across different environments, from development laptops to production servers. However, as applications grow, managing many containers becomes challenging.
+
+This is where orchestration tools come in. They help automate the deployment, scaling, and management of containerized applications. Key benefits include:
+
+- Automating application maintenance
+- Automatically replacing failed containers
+- Managing updates and configurations
+
+Two popular orchestration tools are [Kubernetes](https://docs.docker.com/guides/deployment-orchestration/kube-deploy/) and [Swarm Mode](https://docs.docker.com/engine/swarm/). 
```

Comment:
This part is, in my opinion, the perfect level of detail for this page. 

Maybe we add a few sentences about what k8s and Swarm mode are, too

Reply:
Done!